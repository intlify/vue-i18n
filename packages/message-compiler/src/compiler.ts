import { CompileOptions } from './options'
import { createParser } from './parser'
import { transform } from './transformer'
import { generate, CodeGenResult } from './generator'
import { assign } from '@intlify/shared'

export function baseCompile(
  source: string,
  options: CompileOptions = {}
): CodeGenResult {
  const assignedOptions = assign({}, options)

  // parse source codes
  const parser = createParser(assignedOptions)
  const ast = parser.parse(source)

  // transform ASTs
  transform(ast, assignedOptions)

  // generate javascript codes
  return generate(ast, assignedOptions)
}
