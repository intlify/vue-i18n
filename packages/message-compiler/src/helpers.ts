export const enum HelperNameMap {
  LIST = 'list',
  NAMED = 'named',
  PLURAL = 'plural',
  LINKED = 'linked',
  MESSAGE = 'message',
  TYPE = 'type',
  INTERPOLATE = 'interpolate',
  NORMALIZE = 'normalize'
}
