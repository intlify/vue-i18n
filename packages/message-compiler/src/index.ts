export * from './options'
export * from './errors'
export * from './helpers'
export * from './location'
export * from './parser'
export * from './compiler'
