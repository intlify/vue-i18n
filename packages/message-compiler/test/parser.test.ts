import { CompileError, ParserOptions } from '../src'
import { createParser } from '../src/parser'

test('parse', () => {
  ;[
    { code: 'hello world', name: 'message' },
    { code: 'hello {name} !', name: 'named' },
    { code: 'hello {_name} !', name: 'named' },
    { code: 'hello {0} !', name: 'list' },
    { code: `hello {'kazupon'} !`, name: 'literal' },
    { code: '@:apples', name: 'linked' },
    { code: '@.lower:{0}', name: 'linked modifier list' },
    { code: 'no apples | one apple  |  too much apples  ', name: 'plural' },
    {
      code: `@.lower:{'no apples'} | {1} apple | {count}ã€€apples`, // eslint-disable-line no-irregular-whitespace
      name: 'plural complex'
    },
    { code: `@.: a`, name: 'linked error' },
    { code: `@.:foo`, name: 'linked modifier error' },
    { code: `@:(foo)`, name: 'linked key paren error' },
    { code: `hi @._upper:{_name} !`, name: 'foo' },
    { code: `@.lower:(foo)`, name: 'linked key paren error with modifier' },
    { code: `@.`, name: 'EOF in linked modifier' },
    { code: `|`, name: 'empty plural' }
  ].forEach(({ name, code }) => {
    const errors: CompileError[] = []
    const options: ParserOptions = {
      onError: err => {
        errors.push({ ...err, message: err.message })
      }
    }
    const parser = createParser(options)
    const ast = parser.parse(code)
    expect(ast).toMatchSnapshot(name || JSON.stringify(code))
    if (errors.length) {
      expect(errors).toMatchSnapshot(`${name || JSON.stringify(code)} errors`)
    }
  })
})
