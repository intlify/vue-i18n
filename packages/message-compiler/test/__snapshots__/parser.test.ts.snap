// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse: EOF in linked modifier 1`] = `
Object {
  "body": Object {
    "end": 2,
    "items": Array [
      Object {
        "end": 2,
        "key": Object {
          "end": 2,
          "loc": Object {
            "end": Object {
              "column": 3,
              "line": 1,
              "offset": 2,
            },
            "start": Object {
              "column": 3,
              "line": 1,
              "offset": 2,
            },
          },
          "start": 2,
          "type": 7,
          "value": "",
        },
        "loc": Object {
          "end": Object {
            "column": 3,
            "line": 1,
            "offset": 2,
          },
          "start": Object {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
        },
        "modifier": Object {
          "end": 1,
          "loc": Object {
            "end": Object {
              "column": 2,
              "line": 1,
              "offset": 1,
            },
            "start": Object {
              "column": 2,
              "line": 1,
              "offset": 1,
            },
          },
          "start": 1,
          "type": 8,
          "value": "",
        },
        "start": 0,
        "type": 6,
      },
    ],
    "loc": Object {
      "end": Object {
        "column": 3,
        "line": 1,
        "offset": 2,
      },
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "start": 0,
    "type": 2,
  },
  "end": 2,
  "loc": Object {
    "end": Object {
      "column": 3,
      "line": 1,
      "offset": 2,
    },
    "source": "@.",
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "start": 0,
  "type": 0,
}
`;

exports[`parse: EOF in linked modifier errors 1`] = `
Array [
  Object {
    "code": 11,
    "domain": "parser",
    "location": Object {
      "end": Object {
        "column": 3,
        "line": 1,
        "offset": 2,
      },
      "start": Object {
        "column": 2,
        "line": 1,
        "offset": 1,
      },
    },
    "message": "Unexpected empty linked modifier",
  },
  Object {
    "code": 13,
    "domain": "parser",
    "location": Object {
      "end": Object {
        "column": 3,
        "line": 1,
        "offset": 2,
      },
      "start": Object {
        "column": 2,
        "line": 1,
        "offset": 1,
      },
    },
    "message": "Unexpected lexical analysis in token: 'EOF'",
  },
  Object {
    "code": 12,
    "domain": "parser",
    "location": Object {
      "end": Object {
        "column": 3,
        "line": 1,
        "offset": 2,
      },
      "start": Object {
        "column": 3,
        "line": 1,
        "offset": 2,
      },
    },
    "message": "Unexpected empty linked key",
  },
]
`;

exports[`parse: empty plural 1`] = `
Object {
  "body": Object {
    "cases": Array [
      Object {
        "end": 0,
        "items": Array [],
        "loc": Object {
          "end": Object {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
          "start": Object {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
        },
        "start": 0,
        "type": 2,
      },
      Object {
        "end": 1,
        "items": Array [],
        "loc": Object {
          "end": Object {
            "column": 2,
            "line": 1,
            "offset": 1,
          },
          "start": Object {
            "column": 2,
            "line": 1,
            "offset": 1,
          },
        },
        "start": 1,
        "type": 2,
      },
    ],
    "end": 1,
    "loc": Object {
      "end": Object {
        "column": 2,
        "line": 1,
        "offset": 1,
      },
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "start": 0,
    "type": 1,
  },
  "end": 1,
  "loc": Object {
    "end": Object {
      "column": 2,
      "line": 1,
      "offset": 1,
    },
    "source": "|",
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "start": 0,
  "type": 0,
}
`;

exports[`parse: empty plural errors 1`] = `
Array [
  Object {
    "code": 10,
    "domain": "parser",
    "location": Object {
      "end": Object {
        "column": 2,
        "line": 1,
        "offset": 1,
      },
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "message": "Plural must have messages",
  },
]
`;

exports[`parse: foo 1`] = `
Object {
  "body": Object {
    "end": 21,
    "items": Array [
      Object {
        "end": 3,
        "loc": Object {
          "end": Object {
            "column": 4,
            "line": 1,
            "offset": 3,
          },
          "start": Object {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
        },
        "start": 0,
        "type": 3,
        "value": "hi ",
      },
      Object {
        "end": 19,
        "key": Object {
          "end": 19,
          "key": "_name",
          "loc": Object {
            "end": Object {
              "column": 20,
              "line": 1,
              "offset": 19,
            },
            "start": Object {
              "column": 13,
              "line": 1,
              "offset": 12,
            },
          },
          "start": 12,
          "type": 4,
        },
        "loc": Object {
          "end": Object {
            "column": 20,
            "line": 1,
            "offset": 19,
          },
          "start": Object {
            "column": 4,
            "line": 1,
            "offset": 3,
          },
        },
        "modifier": Object {
          "end": 11,
          "loc": Object {
            "end": Object {
              "column": 12,
              "line": 1,
              "offset": 11,
            },
            "start": Object {
              "column": 5,
              "line": 1,
              "offset": 4,
            },
          },
          "start": 4,
          "type": 8,
          "value": "_upper",
        },
        "start": 3,
        "type": 6,
      },
      Object {
        "end": 21,
        "loc": Object {
          "end": Object {
            "column": 22,
            "line": 1,
            "offset": 21,
          },
          "start": Object {
            "column": 20,
            "line": 1,
            "offset": 19,
          },
        },
        "start": 19,
        "type": 3,
        "value": " !",
      },
    ],
    "loc": Object {
      "end": Object {
        "column": 22,
        "line": 1,
        "offset": 21,
      },
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "start": 0,
    "type": 2,
  },
  "end": 21,
  "loc": Object {
    "end": Object {
      "column": 22,
      "line": 1,
      "offset": 21,
    },
    "source": "hi @._upper:{_name} !",
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "start": 0,
  "type": 0,
}
`;

exports[`parse: linked 1`] = `
Object {
  "body": Object {
    "end": 8,
    "items": Array [
      Object {
        "end": 8,
        "key": Object {
          "end": 8,
          "loc": Object {
            "end": Object {
              "column": 9,
              "line": 1,
              "offset": 8,
            },
            "start": Object {
              "column": 3,
              "line": 1,
              "offset": 2,
            },
          },
          "start": 2,
          "type": 7,
          "value": "apples",
        },
        "loc": Object {
          "end": Object {
            "column": 9,
            "line": 1,
            "offset": 8,
          },
          "start": Object {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
        },
        "start": 0,
        "type": 6,
      },
    ],
    "loc": Object {
      "end": Object {
        "column": 9,
        "line": 1,
        "offset": 8,
      },
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "start": 0,
    "type": 2,
  },
  "end": 8,
  "loc": Object {
    "end": Object {
      "column": 9,
      "line": 1,
      "offset": 8,
    },
    "source": "@:apples",
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "start": 0,
  "type": 0,
}
`;

exports[`parse: linked error 1`] = `
Object {
  "body": Object {
    "end": 5,
    "items": Array [
      Object {
        "end": 3,
        "key": Object {
          "end": 3,
          "loc": Object {
            "end": Object {
              "column": 4,
              "line": 1,
              "offset": 3,
            },
            "start": Object {
              "column": 4,
              "line": 1,
              "offset": 3,
            },
          },
          "start": 3,
          "type": 7,
          "value": "",
        },
        "loc": Object {
          "end": Object {
            "column": 4,
            "line": 1,
            "offset": 3,
          },
          "start": Object {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
        },
        "modifier": Object {
          "end": 1,
          "loc": Object {
            "end": Object {
              "column": 2,
              "line": 1,
              "offset": 1,
            },
            "start": Object {
              "column": 2,
              "line": 1,
              "offset": 1,
            },
          },
          "start": 1,
          "type": 8,
          "value": "",
        },
        "start": 0,
        "type": 6,
      },
      Object {
        "end": 5,
        "loc": Object {
          "end": Object {
            "column": 6,
            "line": 1,
            "offset": 5,
          },
          "start": Object {
            "column": 4,
            "line": 1,
            "offset": 3,
          },
        },
        "start": 3,
        "type": 3,
        "value": " a",
      },
    ],
    "loc": Object {
      "end": Object {
        "column": 6,
        "line": 1,
        "offset": 5,
      },
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "start": 0,
    "type": 2,
  },
  "end": 5,
  "loc": Object {
    "end": Object {
      "column": 6,
      "line": 1,
      "offset": 5,
    },
    "source": "@.: a",
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "start": 0,
  "type": 0,
}
`;

exports[`parse: linked error errors 1`] = `
Array [
  Object {
    "code": 11,
    "domain": "parser",
    "location": Object {
      "end": Object {
        "column": 4,
        "line": 1,
        "offset": 3,
      },
      "start": Object {
        "column": 2,
        "line": 1,
        "offset": 1,
      },
    },
    "message": "Unexpected empty linked modifier",
  },
  Object {
    "code": 9,
    "domain": "tokenizer",
    "location": Object {
      "end": Object {
        "column": 4,
        "line": 1,
        "offset": 3,
      },
      "start": Object {
        "column": 4,
        "line": 1,
        "offset": 3,
      },
    },
    "message": "Invalid linked format",
  },
  Object {
    "code": 12,
    "domain": "parser",
    "location": Object {
      "end": Object {
        "column": 6,
        "line": 1,
        "offset": 5,
      },
      "start": Object {
        "column": 3,
        "line": 1,
        "offset": 2,
      },
    },
    "message": "Unexpected empty linked key",
  },
]
`;

exports[`parse: linked key paren error 1`] = `
Object {
  "body": Object {
    "end": 7,
    "items": Array [
      Object {
        "end": 2,
        "key": Object {
          "end": 2,
          "loc": Object {
            "end": Object {
              "column": 3,
              "line": 1,
              "offset": 2,
            },
            "start": Object {
              "column": 3,
              "line": 1,
              "offset": 2,
            },
          },
          "start": 2,
          "type": 7,
          "value": "",
        },
        "loc": Object {
          "end": Object {
            "column": 3,
            "line": 1,
            "offset": 2,
          },
          "start": Object {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
        },
        "start": 0,
        "type": 6,
      },
      Object {
        "end": 7,
        "loc": Object {
          "end": Object {
            "column": 8,
            "line": 1,
            "offset": 7,
          },
          "start": Object {
            "column": 3,
            "line": 1,
            "offset": 2,
          },
        },
        "start": 2,
        "type": 3,
        "value": "(foo)",
      },
    ],
    "loc": Object {
      "end": Object {
        "column": 8,
        "line": 1,
        "offset": 7,
      },
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "start": 0,
    "type": 2,
  },
  "end": 7,
  "loc": Object {
    "end": Object {
      "column": 8,
      "line": 1,
      "offset": 7,
    },
    "source": "@:(foo)",
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "start": 0,
  "type": 0,
}
`;

exports[`parse: linked key paren error errors 1`] = `
Array [
  Object {
    "code": 12,
    "domain": "parser",
    "location": Object {
      "end": Object {
        "column": 8,
        "line": 1,
        "offset": 7,
      },
      "start": Object {
        "column": 2,
        "line": 1,
        "offset": 1,
      },
    },
    "message": "Unexpected empty linked key",
  },
]
`;

exports[`parse: linked key paren error with modifier 1`] = `
Object {
  "body": Object {
    "end": 13,
    "items": Array [
      Object {
        "end": 8,
        "key": Object {
          "end": 8,
          "loc": Object {
            "end": Object {
              "column": 9,
              "line": 1,
              "offset": 8,
            },
            "start": Object {
              "column": 9,
              "line": 1,
              "offset": 8,
            },
          },
          "start": 8,
          "type": 7,
          "value": "",
        },
        "loc": Object {
          "end": Object {
            "column": 9,
            "line": 1,
            "offset": 8,
          },
          "start": Object {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
        },
        "modifier": Object {
          "end": 7,
          "loc": Object {
            "end": Object {
              "column": 8,
              "line": 1,
              "offset": 7,
            },
            "start": Object {
              "column": 2,
              "line": 1,
              "offset": 1,
            },
          },
          "start": 1,
          "type": 8,
          "value": "lower",
        },
        "start": 0,
        "type": 6,
      },
      Object {
        "end": 13,
        "loc": Object {
          "end": Object {
            "column": 14,
            "line": 1,
            "offset": 13,
          },
          "start": Object {
            "column": 9,
            "line": 1,
            "offset": 8,
          },
        },
        "start": 8,
        "type": 3,
        "value": "(foo)",
      },
    ],
    "loc": Object {
      "end": Object {
        "column": 14,
        "line": 1,
        "offset": 13,
      },
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "start": 0,
    "type": 2,
  },
  "end": 13,
  "loc": Object {
    "end": Object {
      "column": 14,
      "line": 1,
      "offset": 13,
    },
    "source": "@.lower:(foo)",
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "start": 0,
  "type": 0,
}
`;

exports[`parse: linked key paren error with modifier errors 1`] = `
Array [
  Object {
    "code": 12,
    "domain": "parser",
    "location": Object {
      "end": Object {
        "column": 14,
        "line": 1,
        "offset": 13,
      },
      "start": Object {
        "column": 8,
        "line": 1,
        "offset": 7,
      },
    },
    "message": "Unexpected empty linked key",
  },
]
`;

exports[`parse: linked modifier error 1`] = `
Object {
  "body": Object {
    "end": 6,
    "items": Array [
      Object {
        "end": 6,
        "key": Object {
          "end": 6,
          "loc": Object {
            "end": Object {
              "column": 7,
              "line": 1,
              "offset": 6,
            },
            "start": Object {
              "column": 4,
              "line": 1,
              "offset": 3,
            },
          },
          "start": 3,
          "type": 7,
          "value": "foo",
        },
        "loc": Object {
          "end": Object {
            "column": 7,
            "line": 1,
            "offset": 6,
          },
          "start": Object {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
        },
        "modifier": Object {
          "end": 1,
          "loc": Object {
            "end": Object {
              "column": 2,
              "line": 1,
              "offset": 1,
            },
            "start": Object {
              "column": 2,
              "line": 1,
              "offset": 1,
            },
          },
          "start": 1,
          "type": 8,
          "value": "",
        },
        "start": 0,
        "type": 6,
      },
    ],
    "loc": Object {
      "end": Object {
        "column": 7,
        "line": 1,
        "offset": 6,
      },
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "start": 0,
    "type": 2,
  },
  "end": 6,
  "loc": Object {
    "end": Object {
      "column": 7,
      "line": 1,
      "offset": 6,
    },
    "source": "@.:foo",
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "start": 0,
  "type": 0,
}
`;

exports[`parse: linked modifier error errors 1`] = `
Array [
  Object {
    "code": 11,
    "domain": "parser",
    "location": Object {
      "end": Object {
        "column": 4,
        "line": 1,
        "offset": 3,
      },
      "start": Object {
        "column": 2,
        "line": 1,
        "offset": 1,
      },
    },
    "message": "Unexpected empty linked modifier",
  },
]
`;

exports[`parse: linked modifier list 1`] = `
Object {
  "body": Object {
    "end": 11,
    "items": Array [
      Object {
        "end": 11,
        "key": Object {
          "end": 11,
          "index": 0,
          "loc": Object {
            "end": Object {
              "column": 12,
              "line": 1,
              "offset": 11,
            },
            "start": Object {
              "column": 9,
              "line": 1,
              "offset": 8,
            },
          },
          "start": 8,
          "type": 5,
        },
        "loc": Object {
          "end": Object {
            "column": 12,
            "line": 1,
            "offset": 11,
          },
          "start": Object {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
        },
        "modifier": Object {
          "end": 7,
          "loc": Object {
            "end": Object {
              "column": 8,
              "line": 1,
              "offset": 7,
            },
            "start": Object {
              "column": 2,
              "line": 1,
              "offset": 1,
            },
          },
          "start": 1,
          "type": 8,
          "value": "lower",
        },
        "start": 0,
        "type": 6,
      },
    ],
    "loc": Object {
      "end": Object {
        "column": 12,
        "line": 1,
        "offset": 11,
      },
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "start": 0,
    "type": 2,
  },
  "end": 11,
  "loc": Object {
    "end": Object {
      "column": 12,
      "line": 1,
      "offset": 11,
    },
    "source": "@.lower:{0}",
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "start": 0,
  "type": 0,
}
`;

exports[`parse: list 1`] = `
Object {
  "body": Object {
    "end": 11,
    "items": Array [
      Object {
        "end": 6,
        "loc": Object {
          "end": Object {
            "column": 7,
            "line": 1,
            "offset": 6,
          },
          "start": Object {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
        },
        "start": 0,
        "type": 3,
        "value": "hello ",
      },
      Object {
        "end": 9,
        "index": 0,
        "loc": Object {
          "end": Object {
            "column": 10,
            "line": 1,
            "offset": 9,
          },
          "start": Object {
            "column": 7,
            "line": 1,
            "offset": 6,
          },
        },
        "start": 6,
        "type": 5,
      },
      Object {
        "end": 11,
        "loc": Object {
          "end": Object {
            "column": 12,
            "line": 1,
            "offset": 11,
          },
          "start": Object {
            "column": 10,
            "line": 1,
            "offset": 9,
          },
        },
        "start": 9,
        "type": 3,
        "value": " !",
      },
    ],
    "loc": Object {
      "end": Object {
        "column": 12,
        "line": 1,
        "offset": 11,
      },
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "start": 0,
    "type": 2,
  },
  "end": 11,
  "loc": Object {
    "end": Object {
      "column": 12,
      "line": 1,
      "offset": 11,
    },
    "source": "hello {0} !",
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "start": 0,
  "type": 0,
}
`;

exports[`parse: literal 1`] = `
Object {
  "body": Object {
    "end": 19,
    "items": Array [
      Object {
        "end": 6,
        "loc": Object {
          "end": Object {
            "column": 7,
            "line": 1,
            "offset": 6,
          },
          "start": Object {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
        },
        "start": 0,
        "type": 3,
        "value": "hello ",
      },
      Object {
        "end": 17,
        "loc": Object {
          "end": Object {
            "column": 18,
            "line": 1,
            "offset": 17,
          },
          "start": Object {
            "column": 7,
            "line": 1,
            "offset": 6,
          },
        },
        "start": 6,
        "type": 9,
        "value": "kazupon",
      },
      Object {
        "end": 19,
        "loc": Object {
          "end": Object {
            "column": 20,
            "line": 1,
            "offset": 19,
          },
          "start": Object {
            "column": 18,
            "line": 1,
            "offset": 17,
          },
        },
        "start": 17,
        "type": 3,
        "value": " !",
      },
    ],
    "loc": Object {
      "end": Object {
        "column": 20,
        "line": 1,
        "offset": 19,
      },
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "start": 0,
    "type": 2,
  },
  "end": 19,
  "loc": Object {
    "end": Object {
      "column": 20,
      "line": 1,
      "offset": 19,
    },
    "source": "hello {'kazupon'} !",
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "start": 0,
  "type": 0,
}
`;

exports[`parse: message 1`] = `
Object {
  "body": Object {
    "end": 11,
    "items": Array [
      Object {
        "end": 11,
        "loc": Object {
          "end": Object {
            "column": 12,
            "line": 1,
            "offset": 11,
          },
          "start": Object {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
        },
        "start": 0,
        "type": 3,
        "value": "hello world",
      },
    ],
    "loc": Object {
      "end": Object {
        "column": 12,
        "line": 1,
        "offset": 11,
      },
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "start": 0,
    "type": 2,
  },
  "end": 11,
  "loc": Object {
    "end": Object {
      "column": 12,
      "line": 1,
      "offset": 11,
    },
    "source": "hello world",
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "start": 0,
  "type": 0,
}
`;

exports[`parse: named 1`] = `
Object {
  "body": Object {
    "end": 14,
    "items": Array [
      Object {
        "end": 6,
        "loc": Object {
          "end": Object {
            "column": 7,
            "line": 1,
            "offset": 6,
          },
          "start": Object {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
        },
        "start": 0,
        "type": 3,
        "value": "hello ",
      },
      Object {
        "end": 12,
        "key": "name",
        "loc": Object {
          "end": Object {
            "column": 13,
            "line": 1,
            "offset": 12,
          },
          "start": Object {
            "column": 7,
            "line": 1,
            "offset": 6,
          },
        },
        "start": 6,
        "type": 4,
      },
      Object {
        "end": 14,
        "loc": Object {
          "end": Object {
            "column": 15,
            "line": 1,
            "offset": 14,
          },
          "start": Object {
            "column": 13,
            "line": 1,
            "offset": 12,
          },
        },
        "start": 12,
        "type": 3,
        "value": " !",
      },
    ],
    "loc": Object {
      "end": Object {
        "column": 15,
        "line": 1,
        "offset": 14,
      },
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "start": 0,
    "type": 2,
  },
  "end": 14,
  "loc": Object {
    "end": Object {
      "column": 15,
      "line": 1,
      "offset": 14,
    },
    "source": "hello {name} !",
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "start": 0,
  "type": 0,
}
`;

exports[`parse: named 2`] = `
Object {
  "body": Object {
    "end": 15,
    "items": Array [
      Object {
        "end": 6,
        "loc": Object {
          "end": Object {
            "column": 7,
            "line": 1,
            "offset": 6,
          },
          "start": Object {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
        },
        "start": 0,
        "type": 3,
        "value": "hello ",
      },
      Object {
        "end": 13,
        "key": "_name",
        "loc": Object {
          "end": Object {
            "column": 14,
            "line": 1,
            "offset": 13,
          },
          "start": Object {
            "column": 7,
            "line": 1,
            "offset": 6,
          },
        },
        "start": 6,
        "type": 4,
      },
      Object {
        "end": 15,
        "loc": Object {
          "end": Object {
            "column": 16,
            "line": 1,
            "offset": 15,
          },
          "start": Object {
            "column": 14,
            "line": 1,
            "offset": 13,
          },
        },
        "start": 13,
        "type": 3,
        "value": " !",
      },
    ],
    "loc": Object {
      "end": Object {
        "column": 16,
        "line": 1,
        "offset": 15,
      },
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "start": 0,
    "type": 2,
  },
  "end": 15,
  "loc": Object {
    "end": Object {
      "column": 16,
      "line": 1,
      "offset": 15,
    },
    "source": "hello {_name} !",
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "start": 0,
  "type": 0,
}
`;

exports[`parse: plural 1`] = `
Object {
  "body": Object {
    "cases": Array [
      Object {
        "end": 0,
        "items": Array [
          Object {
            "end": 9,
            "loc": Object {
              "end": Object {
                "column": 10,
                "line": 1,
                "offset": 9,
              },
              "start": Object {
                "column": 1,
                "line": 1,
                "offset": 0,
              },
            },
            "start": 0,
            "type": 3,
            "value": "no apples",
          },
        ],
        "loc": Object {
          "end": Object {
            "column": 10,
            "line": 1,
            "offset": 9,
          },
          "start": Object {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
        },
        "start": 0,
        "type": 2,
      },
      Object {
        "end": 12,
        "items": Array [
          Object {
            "end": 21,
            "loc": Object {
              "end": Object {
                "column": 22,
                "line": 1,
                "offset": 21,
              },
              "start": Object {
                "column": 13,
                "line": 1,
                "offset": 12,
              },
            },
            "start": 12,
            "type": 3,
            "value": "one apple",
          },
        ],
        "loc": Object {
          "end": Object {
            "column": 22,
            "line": 1,
            "offset": 21,
          },
          "start": Object {
            "column": 13,
            "line": 1,
            "offset": 12,
          },
        },
        "start": 12,
        "type": 2,
      },
      Object {
        "end": 43,
        "items": Array [
          Object {
            "end": 43,
            "loc": Object {
              "end": Object {
                "column": 44,
                "line": 1,
                "offset": 43,
              },
              "start": Object {
                "column": 27,
                "line": 1,
                "offset": 26,
              },
            },
            "start": 26,
            "type": 3,
            "value": "too much apples  ",
          },
        ],
        "loc": Object {
          "end": Object {
            "column": 44,
            "line": 1,
            "offset": 43,
          },
          "start": Object {
            "column": 27,
            "line": 1,
            "offset": 26,
          },
        },
        "start": 26,
        "type": 2,
      },
    ],
    "end": 43,
    "loc": Object {
      "end": Object {
        "column": 44,
        "line": 1,
        "offset": 43,
      },
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "start": 0,
    "type": 1,
  },
  "end": 43,
  "loc": Object {
    "end": Object {
      "column": 44,
      "line": 1,
      "offset": 43,
    },
    "source": "no apples | one apple  |  too much apples  ",
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "start": 0,
  "type": 0,
}
`;

exports[`parse: plural complex 1`] = `
Object {
  "body": Object {
    "cases": Array [
      Object {
        "end": 20,
        "items": Array [
          Object {
            "end": 21,
            "key": Object {
              "end": 21,
              "loc": Object {
                "end": Object {
                  "column": 22,
                  "line": 1,
                  "offset": 21,
                },
                "start": Object {
                  "column": 9,
                  "line": 1,
                  "offset": 8,
                },
              },
              "start": 8,
              "type": 9,
              "value": "no apples",
            },
            "loc": Object {
              "end": Object {
                "column": 22,
                "line": 1,
                "offset": 21,
              },
              "start": Object {
                "column": 1,
                "line": 1,
                "offset": 0,
              },
            },
            "modifier": Object {
              "end": 7,
              "loc": Object {
                "end": Object {
                  "column": 8,
                  "line": 1,
                  "offset": 7,
                },
                "start": Object {
                  "column": 2,
                  "line": 1,
                  "offset": 1,
                },
              },
              "start": 1,
              "type": 8,
              "value": "lower",
            },
            "start": 0,
            "type": 6,
          },
        ],
        "loc": Object {
          "end": Object {
            "column": 22,
            "line": 1,
            "offset": 21,
          },
          "start": Object {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
        },
        "start": 0,
        "type": 2,
      },
      Object {
        "end": 27,
        "items": Array [
          Object {
            "end": 27,
            "index": 1,
            "loc": Object {
              "end": Object {
                "column": 28,
                "line": 1,
                "offset": 27,
              },
              "start": Object {
                "column": 25,
                "line": 1,
                "offset": 24,
              },
            },
            "start": 24,
            "type": 5,
          },
          Object {
            "end": 33,
            "loc": Object {
              "end": Object {
                "column": 34,
                "line": 1,
                "offset": 33,
              },
              "start": Object {
                "column": 28,
                "line": 1,
                "offset": 27,
              },
            },
            "start": 27,
            "type": 3,
            "value": " apple",
          },
        ],
        "loc": Object {
          "end": Object {
            "column": 34,
            "line": 1,
            "offset": 33,
          },
          "start": Object {
            "column": 25,
            "line": 1,
            "offset": 24,
          },
        },
        "start": 24,
        "type": 2,
      },
      Object {
        "end": 50,
        "items": Array [
          Object {
            "end": 43,
            "key": "count",
            "loc": Object {
              "end": Object {
                "column": 44,
                "line": 1,
                "offset": 43,
              },
              "start": Object {
                "column": 37,
                "line": 1,
                "offset": 36,
              },
            },
            "start": 36,
            "type": 4,
          },
          Object {
            "end": 50,
            "loc": Object {
              "end": Object {
                "column": 51,
                "line": 1,
                "offset": 50,
              },
              "start": Object {
                "column": 44,
                "line": 1,
                "offset": 43,
              },
            },
            "start": 43,
            "type": 3,
            "value": "　apples",
          },
        ],
        "loc": Object {
          "end": Object {
            "column": 51,
            "line": 1,
            "offset": 50,
          },
          "start": Object {
            "column": 37,
            "line": 1,
            "offset": 36,
          },
        },
        "start": 36,
        "type": 2,
      },
    ],
    "end": 50,
    "loc": Object {
      "end": Object {
        "column": 51,
        "line": 1,
        "offset": 50,
      },
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "start": 0,
    "type": 1,
  },
  "end": 50,
  "loc": Object {
    "end": Object {
      "column": 51,
      "line": 1,
      "offset": 50,
    },
    "source": "@.lower:{'no apples'} | {1} apple | {count}　apples",
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "start": 0,
  "type": 0,
}
`;
